### PR TITLE
fix mutually exclusive mapping start pose (#319)

### PR DESCRIPTION
**Pull request by eliasdc**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #319  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | custom |

---

## Description of contribution in a few bullet points

* fixed mutually exclusive `map_start_pose` and `map_start_at_dock`
* if none are set use origin and log an error message

### CODE REVIEW

File: src/slam_toolbox_common.cpp [hunk 1]
```diff
@@ -319,8 +319,15 @@ bool SlamToolbox::shouldStartWithPoseGraph(
         pose.y = read_pose[1];
         pose.theta = read_pose[2];
       }
-    } else {
+    } else if (this->get_parameter("map_start_at_dock", start_at_dock)) {
       start_at_dock = this->get_parameter("map_start_at_dock").as_bool();
+    } else {
+      start_at_dock = false;
```
Comment (SteveMacenski) [inline src/slam_toolbox_common.cpp]: Remove this, its not robust, if you didn't provide either of these parameters, it should fail and the function return false.

File: src/slam_toolbox_common.cpp [hunk 2]
```diff
@@ -319,8 +319,15 @@ bool SlamToolbox::shouldStartWithPoseGraph(
         pose.y = read_pose[1];
         pose.theta = read_pose[2];
       }
-    } else {
+    } else if (this->get_parameter("map_start_at_dock", start_at_dock)) {
```
Comment (SteveMacenski) [inline src/slam_toolbox_common.cpp]: You already set start at dock in this line. The second line here is redundant. Perhaps there's a better way to do this

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: src/slam_toolbox_common.cpp (hunk 1 above)
Comment (eliasdc) [reply src/slam_toolbox_common.cpp]: This would also imply you want to fail if the number of parameters is defined wrong instead of using origin?

Comment (eliasdc) [reply src/slam_toolbox_common.cpp]: Wouldn't it be better to throw an exception when configured badly instead of just returning false?

Comment (SteveMacenski) [reply src/slam_toolbox_common.cpp]: No, if neither are set then its just a normal bring up into mapping mode

File: src/slam_toolbox_common.cpp [hunk 3]
```diff
@@ -300,12 +300,13 @@ bool SlamToolbox::shouldStartWithPoseGraph(
 {
   // if given a map to load at run time, do it.
   this->declare_parameter("map_file_name", std::string(""));
-  this->declare_parameter("map_start_pose", std::vector<double>());
-  this->declare_parameter("map_start_at_dock", false);
+  auto read_pose_param = this->declare_parameter("map_start_pose");
```
Comment (SteveMacenski) [inline src/slam_toolbox_common.cpp]: Please rename these 2 params to `map_start_pose` and `start_at_doc`. Adding a `_param` is redundant. All objects are basically params so its not really meaningful

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
LGTM, just need to clean up variable names

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Thanks! I reviewed from ROS1 and looks like I just messed this up in porting

**Pull request merged**